### PR TITLE
Get snakemake ready to use

### PR DESCRIPTION
**Pull request by IsleZhu**
This branch modified the paths of snakemake output data, to make it consistent with the project. Also will add `--config organism` to allow preparing data by organisms.

### CODE REVIEW

**Comment by IsleZhu** [comment]
The new command `add-sam-tag` works like `guidescanpy add-sam-tag sacCer3.bam.sorted --ce -o output1.sam --output-format sam`. Input and output files can be either `sam` or `bam` format.

File: snakemake/Snakefile [hunk 1]
```diff
@@ -116,21 +117,31 @@ rule guidescan_index:
...
-        guidescan enumerate {params.index_file} --kmers-file {input[1]} --output {output}
-    """
+
+rule add_sam_ce:
+    input: rules.guidescan_enumerate.output
+    output: f"{OUTPUT_DIR}/databases/{{enzyme}}/{{organism}}.bam.sorted"
+    shell:
+        """
+        guidescanpy add-sam-tag {input} --ce -o {output}
+        python -c "import pysam; pysam.sort('-o', '{output}', '{output}'); pysam.index('{output}')"
```
Comment (vineetbansal) [inline snakemake/Snakefile:145]: Let's create an `{{organism}}.bam` first, and then sort it and save as `{{organism}}.bam.sorted`

File: snakemake/config.json [hunk 2]
```diff
@@ -1,9 +1,43 @@
 {
-  "output_dir": "data",
-
+  "output_dir": "../data",
+  "pams": {
+    "cas9": "NGG",
+    "cpf1": "TTTN"
```
Comment (vineetbansal) [inline snakemake/config.json:5]: Let's make an issue on whether this should be `TTTN` or `TTTV` and clarify with Yuri/Henri.

File: src/guidescanpy/commands/add_sam_tag.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    return parser
+
+
+def add_tag(tag_name, input_sam_file, output_sam_file, is_bam):
+    if is_bam:
+        writing_mode = "wb"
+    else:
+        writing_mode = "w"
+
+    with pysam.AlignmentFile(input_sam_file, "r") as input_file, pysam.AlignmentFile(
```
Comment (vineetbansal) [inline src/guidescanpy/commands/add_sam_tag.py:33]: Make sure that "r" works for input BAM files and we don't need "rb"

File: snakemake/config.json [hunk 4]
```diff
@@ -1,9 +1,43 @@
 {
-  "output_dir": "data",
-
+  "output_dir": "../data",
```
Comment (vineetbansal) [inline snakemake/config.json:2]: We should probably keep this as `./data` (make the workflow not modify anything thats outside this folder)

File: snakemake/config.json [hunk 5]
```diff
@@ -1,9 +1,43 @@
...
+  "pams": {
+    "cas9": "NGG",
+    "cpf1": "TTTN"
+  },
+  "organisms": ["sacCer3"],
   "sacCer3": {
     "fna": "https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.fna.gz",
     "gtf": "https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.gtf.gz",
     "chr2acc": "https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_assembly_structure/Primary_Assembly/assembled_chromosomes/chr2acc"
+  },
```
Comment (vineetbansal) [inline snakemake/config.json:12]: Note to @vineetbansal - provide real URLs here.

File: src/guidescanpy/commands/add_sam_tag.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        output_sam_file, writing_mode, header=input_file.header
+    ) as output_file:
+        for read in input_file:
+            match tag_name:
+                case "ce":
+                    tag_value = get_ce_value(read)
+                case _:
+                    raise ValueError(
```
Comment (vineetbansal) [inline src/guidescanpy/commands/add_sam_tag.py:41]: get rid of this exception, bu have `get_tag_value` raise it if it doesn't recognize the tag.

File: src/guidescanpy/commands/add_sam_tag.py (hunk 3 above)
Comment (IsleZhu) [reply src/guidescanpy/commands/add_sam_tag.py:33]: Somehow 'r' works for both `.bam` files and `.sam` files in my environment, but it's definetely better to use `rb` for `bam` to make it less confusing.

The official documentation says:
>If mode is not specified, the method will try to auto-detect in the order ‘rb’, ‘r’, thus both the following should work:
>```
>f1 = pysam.AlignmentFile('ex1.bam')
>f2 = pysam.AlignmentFile('ex1.sam')
>```

Comment (IsleZhu) [reply src/guidescanpy/commands/add_sam_tag.py]: Actually it's better to leave the reading mode blank. pysam's auto-detection is more accurate than our get_format function.

**Review by vineetbansal — APPROVED** [review]

File: snakemake/config.json (hunk 2 above)
Comment (vineetbansal) [reply snakemake/config.json:5]: Ok - looks like we need to stick with `TTTN`. Closing this.

**Pull request merged**